include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,394 @@
+//
+// Copyright (c) 2017 Damian Jarek (damian dot jarek93 at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/djarek/gabia
+//
+
+#ifndef GABIA_SECURE_SOCKET_HPP
+#define GABIA_SECURE_SOCKET_HPP
+
+#include <gabia/bytes.hpp>
+#include <gabia/crypto/aead.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/beast.hpp>
+#include <boost/endian/arithmetic.hpp>
+
+#define CORO_YIELD BOOST_ASIO_CORO_YIELD
+#define CORO_REENTER BOOST_ASIO_CORO_REENTER
+
+namespace gabia {
+
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+
+template <typename DynamicBuffer>
+auto prepare_or_error(DynamicBuffer& buffer, size_t n,
+                      beast::error_code& error) {
+    boost::optional<typename DynamicBuffer::mutable_buffers_type> sequence;
+    try {
+        sequence = buffer.prepare(n);
+    } catch (const std::length_error&) {
+        error = asio::error::message_size;
+    }
+    return sequence;
+}
+
+template <typename Stream>
+class secure_socket {
+public:
+    using next_layer_type = typename std::remove_reference<Stream>::type;
+    using lowest_layer_type =
+        typename beast::get_lowest_layer<next_layer_type>::type;
+    constexpr static auto max_payload_length = 1024;
```

Comment:
Hmm, I have a hard time imagining a use case for this, so it should probably be private.